packages/edit-navigation/src/utils/fetch-link-suggestions.js

```diff
@@ -0,0 +1,74 @@
+/**
+ * WordPress dependencies
+ */
+import apiFetch from '@wordpress/api-fetch';
+import { addQueryArgs } from '@wordpress/url';
+/**
+ * External dependencies
+ */
+import { flatten, slice, map, flow } from 'lodash';
+import { decodeEntities } from '@wordpress/html-entities';
+import { __ } from '@wordpress/i18n';
+
+const mapC = ( f ) => ( arr ) => map( arr, f );
+const toSuggestionLinkObj = ( { id, url, title, subType, type } ) => ( {
+	id,
+	url,
+	title: decodeEntities( title ) || __( '(no title)' ),
+	type: subType || type,
+} );
+/**
+ * Fetches link suggestions from the API. This function is an exact copy of a function found at:
+ *
+ * packages/editor/src/components/provider/index.js
+ *
+ * It seems like there is no suitable package to import this from. Ideally it would be either part of core-data.
+ * Until we refactor it, just copying the code is the simplest solution.
+ *
+ * @param {string} search
+ * @param {Object} [searchArguments]
+ * @param {number} [searchArguments.isInitialSuggestions]
+ * @param {number} [searchArguments.type]
+ * @param {number} [searchArguments.subtype]
+ * @param {Object} [editorSettings]
+ * @param {boolean} [editorSettings.disablePostFormats=false]
+ * @return {Promise<Object[]>} List of suggestions
+ */
+export default (
+	search,
+	{ isInitialSuggestions, type, subtype } = {},
+	{ disablePostFormats = false } = {}
+) => {
+	const perPage = isInitialSuggestions ? 3 : 20;
+
+	const linkTypes = [ 'post', 'term' ];
+	if ( ! disablePostFormats ) linkTypes.push( 'post-format' );
+
+	const typesOfLinksToFetch = type ? [ type ] : linkTypes;
+
+	const toApiFetchBody = ( linkType ) => ( {
+		path: addQueryArgs( '/wp/v2/search', {
+			search,
+			per_page: perPage,
+			type: linkType,
+			subtype,
+		} ),
+	} );
+
+	const toReqBody = flow( [
+		toApiFetchBody,
+		apiFetch,
+		( promise ) => promise.catch( () => [] ),
+	] );
+	const toSuggestionLinksObjs = flow( [
+		flatten,
+		( x ) => slice( x, 0, perPage ),
+		toSuggestionLinkObj,
+	] );
+
+	return flow( [
+		mapC( toReqBody ),
+		mapC( async ( x ) => await x ),
+		mapC( toSuggestionLinksObjs ),
+	] )( typesOfLinksToFetch );
```

Comment:
This functional style is probably going to be a bit divisive among those in the WordPress community. I don't mind it, I definitely enjoy when code has a functional leaning, but I think the toughest part here is the readability, particularly when consider some contributors are going to be completely unfamiliar with this style.

WordPress has always championed a readable style, that's the main reason there are all the extra spaces everywhere. Whether we like it or not, the project's style takes precedence over an individual's.

I think there are ways to mitigate this, though. Write comments to help readers understand what each bit of code is doing. Give meaningful names to variables, even if it makes things a bit longer. Extract some code out to functions that have helpful names.

I think it's worth also considering attending the Core JS meetings and discussing functional programming there and the value of it. In a big open source ecosystem like WordPress, it does take some conversation to start turning heads towards something like functional programming.

Reply:
>  think it's worth also considering attending the Core JS meetings and discussing functional programming there and the value of it.

+1 to this. Also I would suggest that we rewrite this in a separate PR, as it's a thing that does not in any way advance the menu name editing.

Reply:
the refactor moved to https://github.com/WordPress/gutenberg/pull/28843